### PR TITLE
Closed Safaricom Mpesa Daraja (Independent Publisher)

### PR DESCRIPTION
**Pull request by muchirimaina**
---
### When submitting a connector, please make sure that you follow the requirements below, otherwise your PR might be rejected. We want to make you have a well-built connector, a smooth certification experience, and your users are happy :) 

If this is your first time submitting to GitHub and you need some help, please sign up for this [session](https://forms.office.com/pages/responsepage.aspx?id=KtIy2vgLW0SOgZbwvQuRaXDXyCl9DkBHq4A2OG7uLpdUMTFJWFFGVUxBNUFZQjZWRUdaOE5BMFkwNS4u). 

- [x] I attest that the connector doesn't exist on the Power Platform today. I've verified by checking the pull requests in GitHub and by searching for the connector on the platform or in the documentation. 
- [x] I attest that the connector works and I verified by deploying and testing all the operations. 
- [x] I attest that I have added detailed descriptions for all operations and parameters in the swagger file.
- [x] I attest that I have added response schemas to my actions, unless the response schema is dynamic. 
- [x] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [x] If this is a certified connector, I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`. If this is an independent publisher connector, I confirm that I am not submitting a connector icon.

If you are an Independent Publisher, you must also attest to the following to ensure a smooth publishing process:
- [x] I have named this PR after the pattern of "Connector Name (Independent Publisher)" ex: HubSpot Marketing (Independent Publisher)
- [x] Within this PR markdown file, I have pasted screenshots that show: 3 unique operations (actions/triggers) working within a Flow. This can be in one flow or part of multiple flows. For each one of those flows, I have pasted in screenshots of the Flow succeeding. 
- [x] Within this PR markdown file, I have pasted in a screenshot from the Test operations section within the Custom Connector UI.
- [x] If the connector uses OAuth, I have provided detailed steps on how to create an app in the readme.md.

### CODE REVIEW

**Comment by johnmuchiriM** [comment]
The test operations were done and all passed.
daraja

**Comment by johnmuchiriM** [comment]
Validation of the connector passed successfully using the paconn tool
daraja-validation

**Comment by johnmuchiriM** [comment]
The stk push on the phone when testing the operations.

**Comment by johnmuchiriM** [comment]
A simple flow of one operation test. "Lipa na Mpesa" 
darajaFlow

**Comment by johnmuchiriM** [comment]
Hi @Amjed-Ayoub,

I would like to request a review of the submitted connector. Also, my colleague and I are working on this connector as a team of two developers.

Could you please advise on how to add my colleague, whose email is [bkaraba14@gmail.com](mailto:bkaraba14@gmail.com), to the communication thread?

Lastly, I have a question regarding creating a pull request. After forking and cloning the repository, and creating a new branch, I noticed that this particular pull request includes some other commits, which I will attach in an image. I'm not sure how these commits ended up in my new pull request. Could you help me figure this out?

Thank you very much.
Screenshot from 2023-04-05 00-38-57

File: independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json [hunk 1]
```diff
@@ -0,0 +1,620 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "title": "daraja",
```
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json:4]: @johnmuchiriM is it possible to capitalize the first letter of the title?

File: independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+    "description": "A connector that exposes daraja mpesa endpoints",
+    "version": "1.0",
+    "contact": 
+    {
+      "name": "John Muchiri, Bernard Karaba",
+      "url": "https://johnmuchirim.github.io/",
+      "email": "jeanfrancaiseharicot@gmail.com",
+      "email": "bkaraba14@gmail.com"
```
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json:12]: @johnmuchiriM please have both emails in one property. please see below.
 "email": "jeanfrancaiseharicot@gmail.com, bkaraba14@gmail.com"

File: independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+    "https"
+  ],
+   "consumes": [
+    "application/json"
+  ],
+  "produces": [
+    "application/json"
+  ],
+  "paths": {
+    "/oauth/v1/generate": {},
```
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json:27]: @johnmuchiriM can you please explain the empty operations? is it possible to remove it?
Please fix all empty paths in the swagger file.

File: independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json [hunk 4]
```diff
@@ -0,0 +1,620 @@
...
+                },
+                "CustomerMessage": {
+                  "type": "string",
+                  "description": "CustomerMessage"
+                }
+              }
+            }
+          }
+        },
+        "summary": "lipa na mpesa",
```
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json:60]: @johnmuchiriM all information within the swagger file should be in English language.

File: independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json [hunk 5]
```diff
@@ -0,0 +1,620 @@
...
+          }
+        ]
+      }
+    },
+    "/mpesa/stkpushquery/v1/query": {
+      "post": {
+        "responses": {
+          "200": {
+            "description": "default",
+            "schema": {}
```
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json:144]: @johnmuchiriM empty schemas are not allowed. please define a proper schema. Please see [here](https://learn.microsoft.com/en-us/connectors/custom-connectors/coding-standards#responses). 
Please fix all empty schemas in the swagger file.

File: independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json [hunk 6]
```diff
@@ -0,0 +1,620 @@
...
+      "oauth2-auth": [
+        "User.Read"
+      ]
+    }
+  ],
+  "tags": [],
+   "x-ms-connector-metadata": [
+    {
+      "propertyName": "Website",
+      "propertyValue": "https://www.developer.safaricom.co.ke/"
```
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json:607]: @johnmuchiriM this should be a valid URL. 
image

File: independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json [hunk 7]
```diff
@@ -0,0 +1,620 @@
...
+    },
+    {
+      "propertyName": "Privacy policy",
+      "propertyValue": "https://www.developer.safaricom.co.ke"
```
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json:611]: @johnmuchiriM this should link to the privacy policy page please.

File: independent-publisher-connectors/Safaricom Mpesa Daraja/apiProperties.json [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        "title": "add auth",
+        "parameters": {
+          "x-ms-apimTemplateParameter.name": "Authentication",
+          "x-ms-apimTemplateParameter.value": "Bearer HNwhBZTObRdSXOGUfGcUA1VnlMsD",
+          "x-ms-apimTemplateParameter.existsAction": "override",
+          "x-ms-apimTemplate-policySection": "Request"
+        }
+      }
+    ],
+    "publisher": "Bernard Karaba"
```
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/Safaricom Mpesa Daraja/apiProperties.json:46]: @johnmuchiriM you can add both publisher here as well. Please see below.
"publisher": "Publisher1, Publisher2"

Please add the stackOwner property. Please see our template [here](https://github.com/microsoft/PowerPlatformConnectors/blob/dev/templates/Independent%20Publisher/apiProperties.json#:~:text=%22stackOwner%22%3A%20%22Name%20of%20company%20or%20organization%20that%20connector%20is%20connecting%20to.%22). 
"stackOwner": "Name of company or organization that connector is connecting to."

**Review by Amjed-Ayoub — CHANGES_REQUESTED** [review]

**Comment by Amjed-Ayoub** [comment]
> Hi @Amjed-Ayoub,
> 
> I would like to request a review of the submitted connector. Also, my colleague and I are working on this connector as a team of two developers.
> 
> Could you please advise on how to add my colleague, whose email is [bkaraba14@gmail.com](mailto:bkaraba14@gmail.com), to the communication thread?
> 
> Lastly, I have a question regarding creating a pull request. After forking and cloning the repository, and creating a new branch, I noticed that this particular pull request includes some other commits, which I will attach in an image. I'm not sure how these commits ended up in my new pull request. Could you help me figure this out?
> 
> Thank you very much. Screenshot from 2023-04-05 00-38-57

@johnmuchiriM I will include @Bkaraba in all comments in this PR.

Please ignore the previous commits. and make sure to update your branch every time you would like to submit a new PR.

**Comment by johnmuchiriM** [comment]
Hi @Amjed-Ayoub,
I have noted all the requested changes and am currently working on them. I will notify you as soon as everything is resolved. I am excited to work on this particular connector and it's a great experience. I am sure my partner is also excited to work on this connector.
Thank you so much.

**Comment by johnmuchiriM** [comment]
Hi @Amjed-Ayoub,
Everything from earlier conversation has been resolved.

**Review by Amjed-Ayoub — CHANGES_REQUESTED** [review]

**Comment by johnmuchiriM** [comment]
Hi @Amjed-Ayoub,
Removed the empty operation.
Regards,
John.

**Comment by Amjed-Ayoub** [comment]
[[certify-connector]]

**Comment by johnmuchiriM** [comment]
Hi @Amjed-Ayoub, 
I'm doing well, thank you for asking. My partner is also doing great. These are great news! Thank you very much for your support throughout the whole process! I'm excited to see the next steps.

File: independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json [hunk 9]
```diff
@@ -0,0 +1,696 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "title": "Daraja",
+    "description": "A connector that exposes daraja mpesa endpoints",
+    "version": "1.0",
+    "contact": {
+      "name": "John Muchiri, Bernard Karaba",
+      "url": "https://johnmuchirim.github.io/",
+      "email": "jeanfrancaiseharicot@gmail.com"
```
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/Safaricom Mpesa Daraja/apiDefinition.swagger.json:10]: @johnmuchiriM and @Bkaraba please notice that we have only one email address here. please confirm?

File: independent-publisher-connectors/Safaricom Mpesa Daraja/apiProperties.json [hunk 10]
```diff
@@ -0,0 +1,49 @@
+{
+  "properties": {
+    "connectionParameters": {
+      "token": {
+        "type": "oauthSetting",
+        "oAuthSettings": {
+          "identityProvider": "oauth2",
+          "clientId": "4b048dda-9dc5-412b-b881-1f6881f882d8",
```
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/Safaricom Mpesa Daraja/apiProperties.json:8]: @johnmuchiriM and @Bkaraba secret look alike values are not allowed. please replace the client ID with a dummy value. You can use something like "Enter your client ID"

File: independent-publisher-connectors/Safaricom Mpesa Daraja/apiProperties.json [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+              "value": "https://daraja.onrender.com/generate"
+            },
+            "refreshUrl": {
+              "value": "https://daraja.onrender.com/generate"
+            }
+          }
+        }
+      }
+    },
+    "iconBrandColor": "#007ee5",
```
Comment (Amjed-Ayoub) [inline independent-publisher-connectors/Safaricom Mpesa Daraja/apiProperties.json:32]: @johnmuchiriM and @Bkaraba for all our Independent Publisher we ask the iconBrandColor to be "#da3b01". can you please change it. please see out template [here](https://github.com/microsoft/PowerPlatformConnectors/blob/dev/templates/Independent%20Publisher/apiProperties.json).

**Review by Amjed-Ayoub — CHANGES_REQUESTED** [review]

**Comment by Amjed-Ayoub** [comment]
Hello @johnmuchiriM and @Bkaraba,
I hope you are doing well.
Congratulations, your pull request is approved. We will move forward with the certification process. We are preparing your connector for production deployment and will queue up for the next deployment schedule. Once your connector onboards the next deployment schedule, it starts to deploy your connector in our production environments that typically takes 3-4 weeks. 
Please let us know if you have any questions.
Thank you very much for working with us.

**Comment by johnmuchiriM** [comment]
Hi @Amjed-Ayoub,
These are great news! We're super excited to see the next phase!! 
Thank you so much!

**Comment by Amjed-Ayoub** [comment]
[[certify-connector]]

**Comment by Amjed-Ayoub** [comment]
Hello @johnmuchiriM ,
I hope you are doing well.
Your connector is in production deployment and queued up for the next deployment schedule. This process typically takes 3-4 weeks. Please note that the documentation for your connector will not be generated until the deployment process is in the final steps.
Please let us know if you have any questions.
Thank you very much for working with us.

**Comment by muchirimaina** [comment]
Hi @Amjed-Ayoub 

The news gives off such a positive vibe as we head into the weekend. Thanks for your support as we work together.

Have a great weekend ahead!

**Comment by muchirimaina** [comment]
Hi @mamurshe & @Amjed-Ayoub,
I hope you're doing well. I'm just checking in to inquire about the connector deployment status. It's been a while since I heard from the team. Any updates would be greatly appreciated.
Thanks and have a great day!

**Comment by SaiChander-MS** [comment]
@johnmuchiriM While we started the deployment process, we see few gaps in your submission like the steps to create AAD application is invalid in readme.md file. Also, the authorizationUrl and host Url in the commit are invalid.
Please address the review comments, so that we can proceed with next steps.

**Comment by vmanoharas** [comment]
Hello @johnmuchiriM, 

Please let's know have you addressed the comments?

**Comment by muchirimaina** [comment]
Hi @vmanoharas,

I have addressed the comments, but I am facing **a permission error** when I try to push to my branch, which I was using before. I would appreciate an extra set of eyes on this. I have **attached the error**, the **Git log showing previous successful commits**, and my **Git config file image**. For the password, I'm using the personal token from developer settings.

Thanks

Screenshot from 2023-09-10 21-16-03
Screenshot from 2023-09-10 21-15-33
Screenshot from 2023-09-10 21-14-33
Screenshot from 2023-09-10 21-49-43

**Comment by vmanoharas** [comment]
Hello @johnmuchiriM, 

Can you please close this PR and submit new PR?

**Comment by muchirimaina** [comment]
Hello @vmanoharas 
That would be great. Let me proceed with that!

**Comment by muchirimaina** [comment]
I closed this branch in order to open a new one for submitting the same connector.

**Pull request closed without merging**